src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/RazorSemanticTokensInfoService.cs

```diff
@@ -118,21 +122,46 @@ internal class RazorSemanticTokensInfoService : IRazorSemanticTokensInfoService
         CancellationToken cancellationToken,
         string? previousResultId = null)
     {
-        // We'll try to call into the mapping service to map to the projected range for us. If that doesn't work,
-        // we'll try to find the minimal range ourselves.
         var generatedDocument = codeDocument.GetCSharpDocument();
-        if (!_documentMappingService.TryMapToGeneratedDocumentRange(generatedDocument, razorRange, out var csharpRange) &&
-            !TryGetMinimalCSharpRange(codeDocument, razorRange, out csharpRange))
+        int[]? csharpResponse = default;
+
+        if (_languageServerFeatureOptions.TrimRangesForSemanticTokens)
         {
-            // There's no C# in the range.
-            return new List<SemanticRange>();
+            if (!TryGetCSharpRanges(codeDocument, razorRange, out var csharpRanges))
+            {
+                // There's no C# in the range.
+                return new List<SemanticRange>();
+            }
+
+            var parameter = new ProvideSemanticTokensRangesParams(textDocumentIdentifier, documentVersion, csharpRanges, correlationId);
+            var response = await _languageServer.SendRequestAsync<ProvideSemanticTokensRangesParams, ProvideSemanticTokensResponse>(
+                CustomMessageNames.RazorProvideSemanticTokensRangesEndpoint,
+                parameter,
+                cancellationToken).ConfigureAwait(false);
+            csharpResponse = GetMatchingCSharpResponse(documentVersion, response);
         }
+        else
+        {
+            // We'll try to call into the mapping service to map to the projected range for us. If that doesn't work,
+            // we'll try to find the minimal range ourselves.
+            if (!_documentMappingService.TryMapToGeneratedDocumentRange(generatedDocument, razorRange, out var csharpRange) &&
+                !TryGetMinimalCSharpRange(codeDocument, razorRange, out csharpRange))
+            {
+                // There's no C# in the range.
+                return new List<SemanticRange>();
+            }
 
-        var csharpResponse = await GetMatchingCSharpResponseAsync(textDocumentIdentifier, documentVersion, csharpRange, correlationId, cancellationToken).ConfigureAwait(false);
+            var parameter = new ProvideSemanticTokensRangeParams(textDocumentIdentifier, documentVersion, csharpRange, correlationId);
+            var response = await _languageServer.SendRequestAsync<ProvideSemanticTokensRangeParams, ProvideSemanticTokensResponse>(
+                CustomMessageNames.RazorProvideSemanticTokensRangeEndpoint,
+                parameter,
+                cancellationToken).ConfigureAwait(false);
+            csharpResponse = GetMatchingCSharpResponse(documentVersion, response);
```

Comment:
I still don't understand why two different messages is a good thing. Compare lines 154-159, with 136-141. They are identical except for two `s`es. The parameter objects have the same shape, except one has a single range, the other has an array.

Would it not be simpler to just have the algorithm be:

1. Compute the C# ranges to ask about
	a. if the feature flag is set, get an array of all individual C# ranges 
	b. if the feature flag is not set, get an array with one element, for the whole C# range
2. Send the array to the CustomMessageTarget
3. Loop through the array and ask Roslyn for tokens
4. Send all of the Roslyn results back from the CustomMessageTarget ot the language server
5. Loop through the results, stitching them together.

Seems to my like that has very little that actually needs to differ with the feature flag, and therefore means a single code path used for more, which means less maintenance, more test coverage, etc. It also seems more future proof, as in future when there is an LSP method to send a single request for multiple C# ranges, only step 3 needs to change, from a loop to a single request.

Reply:
Note that this approach will mean changing this model in VS Code to have an array, but that seems like a very small price to pay, to me.
https://github.com/dotnet/vscode-csharp/blob/68ef16361bd3643fab17f3cbb0f21a421b221b89/src/razor/src/semantic/serializableSemanticTokensParams.ts#L11

Reply:
I agree with @davidwengier's points here. This doesn't seem to warrant an additional endpoint.